### DISCUSSION TITLE
创建 DTO 的偷懒操作

### DISCUSSION THREAD

**Discussion by wjw-gavin** [midwayjs/midway#3945] [General / 闲聊]
在业务中大部分都有 `CreateDTO` 和 `UpdateDTO`，如果写了 `CreateDTO`，有没有什么方法使 `UpdateDTO` 继承 `CreateDTO` 的所有属性，并使其变为可选属性？
我写了一个 `PartialType` 方法继承

<img width="526" alt="image" src="https://github.com/midwayjs/midway/assets/19986739/1df267cb-9522-417e-944b-a58bca3db485">
<img width="334" alt="image" src="https://github.com/midwayjs/midway/assets/19986739/83823ef5-d9c4-4eb5-96d3-2c0817156005">

但是虽说是属性可选了，但是 `@Rule` 其实还在生效，update 时会校验不通过。
```ts
export function PartialType<T extends new (...args: any[]) => any>(
  classRef: T
): new () => Partial<InstanceType<T>> {
  return class extends classRef {
    constructor(...args: any[]) {
      super(...args);
    }
  };
}
```
请教下大佬怎么修改，或者你们平时都是怎么写的o(╥﹏╥)o

**Comment by waitingsong** [comment]
在 `PartialType` 方法内读取出 Rule 在它后面加上可选属性（ optional() ? )

**Reply by wjw-gavin** [reply]
是的，一开始也打算这么搞呢，不太熟怎么读取呢，用 `MetaData` 吗?

**Reply by waitingsong** [reply]
我觉得应该是可行的。用 midway 提供的 meta 相关方法来操作 https://midwayjs.org/docs/custom_decorator

**Reply by wjw-gavin** [reply]
@waitingsong  大佬能不能帮写下，不太会啊😭

**Reply by waitingsong** [reply]
> @waitingsong 大佬能不能帮写下，不太会啊😭

晚点弄弄

**Comment by waitingsong** [comment]
```ts
import { OptionalDto, RequiredDto } from '@mwcp/share'
```

- `OptionalDto()` 方法用于给所有字段的**字段属性**以及**校验参数**追加**可选**属性，可以同时设定忽略/提取字段名 [DEMO](https://github.com/waitingsong/midway-components/blob/main/packages/share/test/util/user.dto.ts#L17)
- `RequiredDto()` 方法用于给所有字段的**字段属性**以及**校验参数**追加**必选**属性，可以同时设定忽略/提取字段名 [DEMO](https://github.com/waitingsong/midway-components/blob/main/packages/share/test/util/user2.dto.ts#L17)

[Coverage 100%](https://app.codecov.io/gh/waitingsong/midway-components/blob/main/packages%2Fshare%2Fsrc%2Futil%2Fdto.util.ts)

@czy88840616 看是否需要给 @midway/validate 包增加这两个方法

**Reply by czy88840616** [reply]
v3 不动了，新功能要往v4走了

**Comment by waitingsong** [comment]
@wjw-gavin 注意，<del>上面两个方法是给字段的**校验参数**增加相应属性，并不会改变字段本身的属性，一楼需求还要在外面再套一层TS内置方法 `Partical<OptionalDto(...)>` 来设置所有字段为可选</del>。或者我看看是否要把这两个 dto 方法的返回类型也相应更新下

上面两个方法的字段可选/必须属性会与校验属性一致，即
- `OptionalDto()` 结果对象所有字段都是**可选**的，所有字段的校验参数包含**可选**
- `RequiredDto()` 结果对象所有字段都是**必选**的，所有字段的校验参数包含**必选**

**Reply by wjw-gavin** [reply]
嗯嗯，感谢提醒，因为忙别的还没试呢。

**Reply by waitingsong** [reply]
> 嗯嗯，感谢提醒，因为忙别的还没试呢。

已更新代码。不需要手动添加 `Partical` 转换方法。测试有什么问题及时告知我修改。
